distributed/utils_test.py

```diff
@@ -1064,15 +1064,40 @@ async def test_foo(scheduler, worker1, worker2, pytest_fixture_a, pytest_fixture
     def _(func):
         if not iscoroutinefunction(func):
             raise RuntimeError("gen_cluster only works for coroutine functions.")
+        # Patch the signature so pytest can inject fixtures
+        orig_sig = inspect.signature(func)
+        args = [None] * (1 + len(nthreads))  # scheduler, *workers
+        if client:
+            args.insert(0, None)
+
+        bound = orig_sig.bind_partial(*args)
+        new_parameters = [
+            p for name, p in orig_sig.parameters.items() if name not in bound.arguments
+        ]
+        requires_tmpdir_factory_fixture = True
+        if "tmp_path_factory" not in orig_sig.parameters:
+            requires_tmpdir_factory_fixture = False
+            new_parameters.append(
+                inspect.Parameter("tmp_path_factory", inspect.Parameter.KEYWORD_ONLY)
+            )
 
         @functools.wraps(func)
         @clean(**clean_kwargs)
         def test_func(*outer_args, **kwargs):
             async def async_fn():
-                result = None
-                with tempfile.TemporaryDirectory() as tmpdir:
+                with contextlib.ExitStack() as exitstack:
```

Comment:
I'm sure there is much more we could do with an exitstack int his function but I'd rather not touch it for now